### PR TITLE
Add `Base.getindex` methods for `DEIntegrator`

### PR DESCRIPTION
**Pull request by AayushSabharwal**
Fixes https://github.com/SciML/ModelingToolkit.jl/issues/1470

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
This needs tests. extend the symbolic indexing tests to use `init` and try these.

**Review by ChrisRackauckas — APPROVED** [review]

**Comment by AayushSabharwal** [comment]
Codecov doesn't like this, it seems.

**Comment by ChrisRackauckas** [comment]
Run the formatter. `using JuliaFormatter; format("path-to-repo")`

**Comment by ChrisRackauckas** [comment]
Looks like real test failures?

**Comment by AayushSabharwal** [comment]
That's weird. They passed when I ran them locally. I'll try and figure out what's up with this.

**Comment by AayushSabharwal** [comment]
It seems that the fields with the same names in `ODEIntegrator` as `ODESolution` don't do the same thing/represent the same quantity, which makes sense. However, I'm at a bit of a loss as to what they actually are.

Additionally, do integrators support _all_ of the indexing capabilities of solutions? Are there some differences?

**Comment by ChrisRackauckas** [comment]
> EDIT: Never mind, I think forwarding to integrator.sol should work

No, don't use that. That assumes someone is saving the solution. In fact, turn saving off in the tests.

> It seems that the fields with the same names in ODEIntegrator as ODESolution don't do the same thing/represent the same quantity, which makes sense. However, I'm at a bit of a loss as to what they actually are.

Integrators are only the current time point, solution is the whole time series. Because of this, indexing an integrator for a time series is undefined. `integrator[i]` is like `sol[i,j]` at the `j'th` time point.

**Comment by AayushSabharwal** [comment]
Ah, got it, thanks! I'll update the tests first then fix the methods.

**Comment by AayushSabharwal** [comment]
What does `sol[:, j]` do when `j` is an integer? I assumed timesteps would be floats

**Comment by AayushSabharwal** [comment]
Why does `integrator(1.0)` work regardless of where `integrator` is in time? If they are only the current time, shouldn't it not work? Additionally, how do I turn of saving, as you mentioned to do?

**Comment by AayushSabharwal** [comment]
Tests won't pass, so no point running them

**Comment by AayushSabharwal** [comment]
The integrator tests pass now (considering I commented most of them since I can't explain why integrator works that way), and a couple come up as broken but it doesn't tell which.

File: test/downstream/integrator_indexing.jl [hunk 1]
```diff
@@ -37,111 +37,100 @@ p = [lorenz1.σ => 10.0,
...
-@test_throws Any integrator['a', [1, 2, 3]]
 
-@test integrator[a] isa AbstractVector
-@test integrator[a, 1] isa Real
-@test integrator[a, 1:5] isa AbstractVector
-@test integrator[a, [1, 2, 3]] isa AbstractVector
+@test integrator[a] isa Real
+# @test integrator[a, 1] isa Real
+# @test integrator[a, 1:5] isa AbstractVector
+# @test integrator[a, [1, 2, 3]] isa AbstractVector
```
Comment (ChrisRackauckas) [inline test/downstream/integrator_indexing.jl:49]: What do these mean? These should just throw.

File: test/downstream/integrator_indexing.jl [hunk 2]
```diff
@@ -37,111 +37,100 @@ p = [lorenz1.σ => 10.0,
...
-@test integrator9.u isa Vector
-@test first(integrator9.u) isa Real
-@test length(integrator9.u) == 11
-@test length(integrator9.t) == 11
-@test collect(integrator9[t]) ≈ integrator3.t
-@test collect(integrator9[t, 1:5]) ≈ integrator3.t[1:5]
-
-integrator10 = integrator(0.1, idxs = 2)
-@test integrator10 isa Real
+# It doesn't because this returns a Vector{Vector{T}} and not a DiffEqArray
```
Comment (ChrisRackauckas) [inline test/downstream/integrator_indexing.jl:59]: Good point, that would be good to solve separately.

Comment (ChrisRackauckas) [reply test/downstream/integrator_indexing.jl:59]: Open an issue?

**Comment by ChrisRackauckas** [comment]
Bring back the 3 that should throw and I think this is good.

File: test/downstream/integrator_indexing.jl (hunk 2 above)
Comment (AayushSabharwal) [reply test/downstream/integrator_indexing.jl:59]: Sure

**Comment by AayushSabharwal** [comment]
Is there anything else left in this PR?

**Pull request merged**